### ISSUE TITLE
关于训练代码的几点疑问

### ISSUE THREAD

**Issue by hyfu0583** [tzco/Diffusion-wo-CFG#6]
@tzco 
作者您好，最近有幸拜读您的文章，感觉想法很棒
对于训练代码我有以下疑问，还望帮忙解答

**1. 关于参数w的值**
    
根据paper所说，w=1.5时效果最好
但根据代码
   ```
   ......
   scale[:num_mg] = torch.rand(num_mg, device=device) * (args.mgw[1] - args.mgw[0]) + args.mgw[0]
   w = torch.where(t[:num_mg] < args.mg_high, scale[:num_mg] - 1, 0)
   ......
   target[:num_mg] = target[:num_mg] + w.view(-1, 1, 1, 1) * (pred_w_cond - pred_wo_cond)
  ```
以及训练代码中默认的参数值args.mgw[1]=args.mgw[0]=1.45,  args.mg_high=0.75
我理解的是t<0.75时，w=0.45 、  t>0.75时w=0
这在paper中似乎没有提及，请问它们的值是如何确定的呢，w的值最后为什么是0.45呢？

**2. 关于训练时样本按λ比例drop 条件（label）**

我理解MG的是将训练target改为ϵ′ = ϵ + w · sg(˜ϵθ(xt, t, c) − ϵ˜θ(xt, t, ∅)). 
即可在推理阶段避免两次调用Dit，且实现与CFG一样的高质量图像。

但在训练时按λ比例drop 条件这一步似乎与CFG训练时是一致的，
只是drop label的样本所占的比例比较小，不知道我理解的有没有问题？
如何理解使用MG-loss的情况下，在训练阶段按比例drop 条件还是能提升合成图像的质量呢？

**Comment by zaptrem** [comment]
Also, how does the automatic adjustment of W work? I can't seem to find it in the code. What metric for quality does it use?

**Comment by tzco** [comment]
Hi hyfu0583,

1. 关于w的值为什么最后是 $1.45-1=0.45$ ，可以参考在另一个issue中的回答（[Issue #7](https://github.com/tzco/Diffusion-wo-CFG/issues/7#issuecomment-2693523230)）。另外，`args.mg_high=0.75`这一参数是参考了[REPA](https://github.com/sihyun-yu/REPA?tab=readme-ov-file#4-evaluation)的实现。这两个参数的值都是在实验中总结出来的。根据经验，它们的值和对应的CFG方法有近似关系。

2. 在MG中，修改训练这一步实际上是使用了贝叶斯公式。要做到贝叶斯估计，我们需要得到 $p(x\_t)$ 的估计，因此我们仿照CFG的做法，通过引入并训练额外的empty label来估计 $p\_{\theta}(x\_t|\varnothing)\approx p(x\_t)$ 。同时，我们也给出了使用全概率公式来估计 $p(x\_t)$ 的方法，但这种做法需要更多采样数才能得到较好的效果。所以我们最后选择了使用empty label，并在训练中按一定比例drop label。

**Comment by tzco** [comment]
> Also, how does the automatic adjustment of W work? I can't seem to find it in the code. What metric for quality does it use?

Thanks for your interest in our work. Currently we have released the main code for training and testing of our SOTA results. We will also release further detailed implementations, such as DiT models, automatic adjustment of w, and 512 resolution supports, in the near future.

**Comment by hyfu0583** [comment]
hi tzco， 感谢您的回复，我大概明白了。
